pkg/policy/selectorcache.go

```diff
@@ -199,9 +244,167 @@ func (l *labelIdentitySelector) updateSelections() {
 	l.setSelections(&selections)
 }
 
-type fqdnSelector struct {
-	labelIdentitySelector
-	// fqdnSelections map[string]identity.NumericIdentity // identity.String(): "cidr:1.1.1.1" -> identity of 1.1.1.1
+func (l *fqdnSelector) setSelections(selections *[]identity.NumericIdentity) {
+	if len(*selections) > 0 {
+		atomic.StorePointer(&l.selections, unsafe.Pointer(selections))
+	} else {
+		atomic.StorePointer(&l.selections, unsafe.Pointer(&emptySelection))
+	}
+}
+
+// updateSelections updates the immutable slice representation of the
+// cached selections after the cached selections have been changed.
+//
+// lock must be held
+func (l *fqdnSelector) updateSelections() {
+	selections := make([]identity.NumericIdentity, len(l.cachedSelections))
+	i := 0
+	for nid := range l.cachedSelections {
+		selections[i] = nid
+		i++
+	}
+	// Sort the numeric identities so that the map iteration order
+	// does not matter. This makes testing easier, but may help
+	// identifying changes easier also otherwise.
+	sort.Slice(selections, func(i, j int) bool {
+		return selections[i] < selections[j]
+	})
+	l.setSelections(&selections)
+}
+
+// UpdateFQDNSelector updates the mapping of fqdnKey (the FQDNSelector from a
+// policy rule as a string) to to the provided list of identities. If the contents
+// of the cachedSelections differ from those in the identities slice, all
+// users are notified.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnKey string, identities []identity.NumericIdentity) {
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+
+	var (
+		fqdnSel *fqdnSelector
+		ok      bool
+	)
+
+	selector, exists := sc.selectors[fqdnKey]
+	if !exists || selector == nil {
+		fqdnSel = &fqdnSelector{
+			key:              fqdnKey,
+			users:            make(map[CachedSelectionUser]struct{}),
+			cachedSelections: make(map[identity.NumericIdentity]struct{}),
+		}
+		sc.selectors[fqdnKey] = fqdnSel
+	} else {
+		fqdnSel, ok = selector.(*fqdnSelector)
+		if !ok {
+			log.WithFields(logrus.Fields{
+				"key":       fqdnKey,
+				"valueType": reflect.TypeOf(fqdnSel),
+			}).Error("key provided to UpdateFQDNSelector does not map to an FQDNSelector")
```

Comment:
Why logging this as an error? How can the user fix it? As a user I don't even know what is the `UpdateFQDNSelector`

Reply:
This would be a problem on the programming side - I will change this to `Fatal` instead since this is something that should never, ever happen.

Reply:
Actually, given that I am changing the parameter for the function to be `api.FQDNSelector` instead of `string`, I can just remove this.